### PR TITLE
Take activity into account in talentpool

### LINKED ISSUE #2302: Nominations: Take into account activity when reviewing
We should check that a user has been active in the past 7 days before choosing them to review.

This will require an endpoint on site that queries metricity for a user's message count.

I don't think we need to worry about clearing out users that stop being active completely in this issue, it might be better to update the `!tp list` command to display activity information somehow so they can be cleared out every now and again.

### PR DESCRIPTION
**Pull request by wookie184**
Closes #2302 

- Only posts review for users who have been active in the past 7 days
- Remove users who have been inactive in the past 30 days from the talent pool
- Make the tp list command sort based on when the review will actually happen, and display
- Send a message in nomination-discussion if a user is automatically removed from the talentpool.

It's not thoroughly tested yet and I will do some more, although I think it's good enough to open and get reviews.

### CODE REVIEW

File: bot/exts/recruitment/talentpool/_cog.py
Comment (shtlrs) [inline bot/exts/recruitment/talentpool/_cog.py:30]: Wouldn't it be a better idea to move these to their own section maybe in the config file ?

File: bot/exts/recruitment/talentpool/_cog.py
Comment (shtlrs) [inline bot/exts/recruitment/talentpool/_cog.py:150]: Wouldn't this fail if the channel is not in cache ?

If that's the case, then we might end up with a null value for that channel .

We could either add our own `get_or_fetch` to bot-core, or have a null-check here.


Danny said that `get_or_fetch` is not there on purpose, here's the [issue](https://github.com/Rapptz/discord.py/issues/9213) for reference.

File: bot/exts/recruitment/talentpool/_cog.py
Comment (shtlrs) [inline bot/exts/recruitment/talentpool/_cog.py:152]: I think that we should reconsider what "active" means here, having it set to any value > 0 is not quite deterministic, as even 1 message would label them as "active", WDYT ?

File: bot/exts/recruitment/talentpool/_cog.py
Comment (shtlrs) [inline bot/exts/recruitment/talentpool/_cog.py:235]: I liked the way we crossed off the name before, it's more catchy that way.
Can't we combine them both ? e.g cross off the name **and** say that they're not part of our server anymore ?

File: bot/exts/recruitment/talentpool/_cog.py
Comment (shtlrs) [inline bot/exts/recruitment/talentpool/_cog.py:503]: Same thing about the `get_channel` here, it might even start to be redudnant :/

**Review by shtlrs — CHANGES_REQUESTED** [review]
This is only a code review, will test it later on.

Other than that, this looks beautiful !

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py:150]: Normal text channels (i.e. not threads) *should* always be cached, it's something we assume quite a lot throughout the bot.

I do agree that this isn't ideal, although i'm not sure what the nicest approach is. At the very least it's annoying to have my editor complain that it could be `None`.

Since it's something that shouldn't be `None`, I don't think we should just ignore if it is (e.g. early return), as this could hide some bugs.

One solution is something i'd thought of when thinking about how to improve type-hintedness (https://github.com/python-discord/bot/issues/1773#issuecomment-1251021265). This would involve creating our own method on `bot` which raises an error if the channel can't be found instead of returning `None`. We could even extend this and provide e.g. a `get_textchannel` method which ensures that the channel is a textchannel and raises an error if it isn't, giving us a more precise type in the rest of the code (otherwise editors will still complain since e.g. voicechannels don't have send).

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py:30]: Yeah, i'm not sure what our general policy on this should be now. I know we wanted to try and keep the config file fairly concise and remove unnecessary things, although making durations easily configurable for development would probably be helpful.

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py:150]: Ah I didn't realise our get_or_fetch channel already never returned `None`, i'll use that https://bot-core.pythondiscord.com/main/output/pydis_core.utils.channel.html#pydis_core.utils.channel.get_or_fetch_channel

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py:30]: @shtlrs do you think it would be easier to do this now, or in a new PR once #2408 is merged to avoid conflicts? I'm happy to do either.

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py]: https://github.com/python-discord/bot/pull/2418/commits/076f0714a43006bff7c5d39aa80f898ab3254c26

File: bot/exts/recruitment/talentpool/_cog.py
Comment (mbaruh) [inline bot/exts/recruitment/talentpool/_cog.py:504]: The implementation of `get_or_fetch_channel` makes me generally uncomfortable since there's no error handling. You'll be bombarded with errors if say something happens to the channel.

**Review by mbaruh — CHANGES_REQUESTED** [review]
Great additions, and some great style improvements as well.

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py:128]: `!tp list` is still one less character ;) I can add it back if you want it but i'd prefer keeping it removed.

FWIW I removed `all` and `watched` because I feel like it's harder to learn the cog commands if see them used with different invocations.

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py:165]: There is one on the cog, although it does it by user ID so would make an unnecessary API call to refetch the nomination in this case. Given it's a single line I think it's probably fine as is.

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py:220]: Given the names of the kwargs it wouldn't make sense to pass both as `True`. 

I can't think of a better way of naming the options without adding e.g. an enum for criteria and order or splitting it into multiple functions, which feels a bit overkill given this works for the current case.

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py:247]: > I'm not sure I like the implication that newer nominations are not ready for review. The min amount of days is a detail for the auto-reviewer. If a moderator has a good reason to put someone for review early, then they should. I'm guessing this isn't what you meant, but it's how it reads once this becomes a user-facing text.

Yeah that's fair

> Regardless, since we're now auto-removing inactive nominations, the activity criterion in is_nomination_ready_for_review seems unnecessary as long as the count in messages_per_user is produced with the same amount of days as the inactivity criterion.

The days value isn't the same though (7d for inactivity and 30d for removal). I think it makes sense for them to be separate too, I don't think we need to review anyone who hasn't been active in the past 7d, but that's definitely too short for complete removal.

> What can be done here (tell what you think) is to instead have two subtitles in the embed: Recent Nominations and Nominations Ready for Auto-Review. This would also help avoid overloading each line with information.

I like this idea, i'll try it out. Do you think we'd need an indication for those without recent activity, (maybe a 💤 emoji or something? Not 100% clear but the best I can think of), or not?

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py:30]: Fine by me, although I don't think it needs to be *too* long. What were you thinking? maybe 40 or 50?

Comment (mbaruh) [reply bot/exts/recruitment/talentpool/_cog.py:30]: Something like that yeah, 45 would work.

File: bot/exts/recruitment/talentpool/_cog.py
Comment (mbaruh) [reply bot/exts/recruitment/talentpool/_cog.py:220]: Hmm, I don't see the issue. You can order by priority, just in reverse.

Comment (mbaruh) [reply bot/exts/recruitment/talentpool/_cog.py:220]: Not a big deal though, since this doesn't affect any planned feature. Your call.

File: bot/exts/recruitment/talentpool/_cog.py
Comment (mbaruh) [reply bot/exts/recruitment/talentpool/_cog.py:247]: > The days value isn't the same though

Ah right

> Do you think we'd need an indication for those without recent activity

A 💤 could work, though that has the issue you mentioned above of using implicit meanings. I guess it will be fine if it's documented in the help embed.

File: bot/exts/recruitment/talentpool/_cog.py
Comment (shtlrs) [reply bot/exts/recruitment/talentpool/_cog.py:30]: I think it's fine to leave as is for now, we also count on waiting for @mbaruh's filters PR to get done so that we could merge.
I would then prune all the filter constants, and try to group the rest by context as  well.

File: bot/exts/recruitment/talentpool/_cog.py
Comment (shtlrs) [reply bot/exts/recruitment/talentpool/_cog.py:159]: @wookie184 I'll resolve my above comment that addresses the same point, and leave the discussion here

File: bot/exts/recruitment/talentpool/_cog.py
Comment (shtlrs) [reply bot/exts/recruitment/talentpool/_cog.py:220]: What about having an `NominationOrdering` enum, that would have the `priority` and `nomination_date` or sthg within those lines ?

File: bot/exts/recruitment/talentpool/_cog.py
Comment (shtlrs) [reply bot/exts/recruitment/talentpool/_cog.py:152]: I see your point.

I guess it's fine to not over engineer this for this particular case. We can always bring it back up if we think it has an impact

File: bot/exts/recruitment/talentpool/_cog.py
Comment (mbaruh) [reply bot/exts/recruitment/talentpool/_cog.py:30]: (my PR isn't blocking, by all means keep the constants there and I'll just resolve the conflict by deleting them)

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py:220]: > Hmm, I don't see the issue. You can order by priority, just in reverse.

Yeah, although the current kwarg name is `oldest_first`, not `in_reverse`, and if I renamed it to `in_reverse` it would be less clear that it was sorting by age by default. Pretty minor, I know, but given it's not something we're using at the moment it seems easier to leave as-is.

File: bot/exts/recruitment/talentpool/_cog.py
Comment (mbaruh) [reply bot/exts/recruitment/talentpool/_cog.py:235]: Given the 💤 idea below, you could use :userleave: from the bot's emojis in this case. Just a thought.

Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py]: I've left it as is for now so it's sort of symetrical with the `({member.name}#{member.discriminator})` that other's have, although i'm happy to change it.

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py]: https://github.com/python-discord/bot/pull/2418/commits/ca40036b17135f2abdcb27e56b950a80b584537e

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py]: Bumped to 45 https://github.com/python-discord/bot/pull/2418/commits/ca40036b17135f2abdcb27e56b950a80b584537e

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py]: Made it kwarg only and updated wording https://github.com/python-discord/bot/pull/2418/commits/e9aa8f1621f9fcff0ace8e6cbe37c98449641773

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py]: I made it three groups, "being reviewed", "recent nominations", and "others by autoreview priority", thoughts?

https://github.com/python-discord/bot/pull/2418/commits/e9aa8f1621f9fcff0ace8e6cbe37c98449641773

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py:160]: Added the duration https://github.com/python-discord/bot/pull/2418/commits/ca40036b17135f2abdcb27e56b950a80b584537e

File: bot/exts/recruitment/talentpool/_cog.py
Comment (mbaruh) [reply bot/exts/recruitment/talentpool/_cog.py]: If that's the case then we no longer need the `(reviewed)` part. Also `recent nominations` and `others by autoreview priority` doesn't tell me what it means for the nomination to be recent (which I think what you were trying to accomplish with the `(ready for review)`). That's why I suggested `Nominations Ready for Auto-Review`, because it implies that recent nominations won't be auto-reviewed.

Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py]: > If that's the case then we no longer need the `(reviewed)` part. 


I made it so it only appears for the newest/oldest subcommands which don't do any grouping.


> Also `recent nominations` and `others by autoreview priority` doesn't tell me what it means for the nomination to be recent (which I think what you were trying to accomplish with the `(ready for review)`).


It's explained in the command help, although the actual conditions are quite complicated so it's hard to get all information across in a simple way.

There's not a specific use for listing nominations, which makes it harder to design a useful output. I wanted to be able to get an idea of who might be put up for review next, I don't really mind exactly how that's done.


> That's why I suggested `Nominations Ready for Auto-Review`, because it implies that recent nominations won't be auto-reviewed.


The "other" section doesn't necessarily mean ready for review though, e.g. if the user is not on the server or doesn't have recent activity.

Comment (mbaruh) [reply bot/exts/recruitment/talentpool/_cog.py]: I just think that once you don't say what it means for a nomination to be recent, it's unclear why it should be in a separate section. It will probably be at the bottom of the priority list anyway.

Either way, it's not a blocker. You could ask the mods what they think a useful output would be.

**Review by shtlrs — CHANGES_REQUESTED** [review]
There's a bug that needs to be fixed, otherwise no one is ready for review :p

Other than that, it works great when i fix it locally :D

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py:28]: I do quite like in-file constants for cog specific stuff, they're really easy to add/change, and only really make sense along with their usage anyway, so it's quite nice to have everything together for quick reference when writing code.

For durations the annoying thing is that it's hard to configure to use a shorter duration for testing in development (so you don't have to wait 45 days to see if it works :P). You have to change it and then change it back before commits.

Overall I don't have a very strong opinion, although I would be happy to leave it here for now.

Comment (shtlrs) [reply bot/exts/recruitment/talentpool/_cog.py:28]: > For durations the annoying thing is that it's hard to configure to use a shorter duration for testing in development (so you don't have to wait 45 days to see if it works :P). You have to change it and then change it back before commits.

That is actually the only thing that I find annoying.

Let's leave them here then, it's not something we got to work on / change / test that frequently anyway.

File: bot/exts/recruitment/talentpool/_cog.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_cog.py:158]: For logging I think we usually use this style across the bot, https://blog.pilosus.org/posts/2020/01/24/python-f-strings-in-logging/ has some reasons and I think there's also potential security issues if you're passing untrusted input (not relevant in this case)

File: bot/exts/recruitment/talentpool/_review.py
Comment (wookie184) [reply bot/exts/recruitment/talentpool/_review.py:149]: Ahh, good find, thanks!

Comment (wookie184) [reply bot/exts/recruitment/talentpool/_review.py]: Fixed

**Review by shtlrs — APPROVED** [review]
:shipit:

**Review by mbaruh — APPROVED** [review]

**Pull request merged**
